src/univariate/continuous/generalizedpareto.jl

```diff
@@ -169,6 +169,189 @@ function quantile(d::GeneralizedPareto{T}, p::Real) where T<:Real
+    return fit(GeneralizedParetoKnownMu(μ), x; kwargs...)
+end
+function StatsBase.fit(g::GeneralizedParetoKnownMu, x::AbstractArray; kwargs...)
+    return fit_empiricalbayes(g, x; kwargs...)
+end
+
+# Note: our ξ is ZhangStephens2009's -k, and our θ is ZhangStephens2009's -θ
+
+function fit_empiricalbayes(
+    g::GeneralizedParetoKnownMu,
+    x::AbstractArray;
+    sorted::Bool=issorted(vec(x)),
+    improved::Bool=true,
+    min_points::Int=30,
+)
+    μ = g.μ
+    T = Base.promote_eltype(x, μ)
+    # fitting is faster when the data are sorted
+    xsorted = sorted ? vec(x) : sort(vec(x))
+    xmin, xmax = @inbounds xsorted[1], xsorted[end]
+    if xmin ≈ xmax
+        # support is nearly a point. solution is not unique; any solution satisfying the
+        # constraints σ/ξ ≈ 0 and ξ < 0 is acceptable. we choose the ξ = -1 solution, i.e.
+        # the uniform distribution
+        return GeneralizedPareto(μ, max(eps(zero(T)), xmax - μ), -one(μ))
+    end
+    # estimate θ using empirical bayes
+    θ_hat = _fit_gpd_θ_empirical_bayes(μ, xsorted, min_points, improved)
+    # estimate remaining parameters using MLE
+    return fit_mle(GeneralizedPareto, xsorted; μ=μ, θ=θ_hat)
+end
+
+# estimate θ̂ = ∫θp(θ|x,μ)dθ, i.e. the posterior mean using quadrature over grid
+# of minimum length `min_points + floor(sqrt(length(x)))` uniformly sampled over an
+# empirical prior
+function _fit_gpd_θ_empirical_bayes(μ, xsorted, min_points, improved)
+    T = Base.promote_eltype(xsorted, μ)
+    n = length(xsorted)
+
+    # empirical prior on y = r/(xmax-μ) + θ
+    θ_prior = if improved
+        _gpd_empirical_prior_improved(μ, xsorted, n)
+    else
+        _gpd_empirical_prior(μ, xsorted, n)
+    end
+
+    # quadrature points uniformly spaced on the quantiles of the θ prior
+    npoints = min_points + floor(Int, sqrt(n))
+    p = ((1:npoints) .- one(T)/2) ./ npoints
+    θ = quantile.(Ref(θ_prior), p)
+
+    # estimate mean θ over the quadrature points
+    # with weights as the normalized profile likelihood 
+    lθ = _gpd_profile_loglikelihood.(μ, θ, Ref(xsorted), n)
+    lθ_norm = logsumexp(lθ)
+    θ_hat = @inbounds sum(1:npoints) do j
+        wⱼ = exp(lθ[j] - lθ_norm)
+        return θ[j] * wⱼ
+    end
+
+    return θ_hat
+end
+
+# Zhang & Stephens, 2009
+function _gpd_empirical_prior(μ, xsorted, n=length(x))
+    xmax = xsorted[n]
+    μ_star = -inv(xmax - μ)
+    x_25 = xsorted[fld(n + 2, 4)]
+    σ_star = inv(6 * (x_25 - μ))
+    ξ_star = 1//2
+    return GeneralizedPareto(μ_star, σ_star, ξ_star)
+end
+
+# Zhang, 2010
+function _gpd_empirical_prior_improved(μ, xsorted, n=length(x))
+    xmax = xsorted[n]
+    μ_star = -inv(xmax - μ) * ((n - 1) // (n + 1))
+    p = (3:9) ./ oftype(μ_star, 10)
+    q = [1 .- p; 1 .- p .^2]
+    xquantiles = if VERSION ≥ v"1.5.0"
+        quantile(xsorted, q; sorted=true, alpha=0, beta=1)
+    else
+        quantile(xsorted, q; sorted=true)
+    end
+    x1mp, x1mp2 = @views xquantiles[1:7], xquantiles[8:14]
```

Comment:
An alternative would be to work with two separate `q1` and `q2` from the start. Since the array is sorted it shouldn't impact performance much (?) if `quantile` is called twice.

As a matter of taste, I prefer separate assignments over the tuple/splatting syntax here:

```suggestion
    x1mp = @views xquantiles[1:7]
    x1mp2 = @views xquantiles[8:14]
```